d2/src/main/java/com/linkedin/d2/balancer/clients/RetryClient.java

```diff
@@ -231,12 +302,34 @@ public void onError(Throwable e)
           }
           else
           {
+            double maxClientRequestRetryRatio;
+            try
+            {
+              Map<String, Object> transportClientProperties =
+                  _balancer.getLoadBalancedServiceProperties(_serviceName).getTransportClientProperties();
```

Comment:
I checked the implementation of getLoadBalancedServiceProperties. It will register a listener on ZK properties event bus only if the service has not already been listened on. And after the listener is registered, subsequent call to getLoadBalancedServiceProperties will directly read the properties from LoadBalancerState, which is a local variable. The RequestTimeoutClient is using the same API to get TransportClientProperties: https://github.com/linkedin/rest.li/blob/99999589f89a168a0528aabc4d25c797b755b57b/d2/src/main/java/com/linkedin/d2/balancer/clients/RequestTimeoutClient.java#L150.

The method without a callback is deprecated because we want it to be async. I changed the impl here to use the callback version.